add security configuration for endpoints

chsegala opened the pull request:
Add Support for `security` constructs.

## Usage:
```
// add security scheme to components
const spec = tsoas.getOpenApiSpec(typeNamesForSecureTests);

        spec.components.securitySchemes = {
            bearerToken: {
                type: "http",
                scheme: "bearer",
                bearerFormat: "JWT",
                description: "Bearer Token with JWT",
            },
            basicAuth: {
                type: "http",
                scheme: "basic",
            },
        };
```

```
// on Apis, add the wanted schemes
type EditBookSecureApi = {
    path: "/category/book/:id";
    method: "PATCH";
    security: [{ bearerToken: []; basicAuth: [] }];
    param: { id: number };
    query: EditBookQuery;
    body: {};
    responses: { "200": Response<EditBookRes> };
};
```

alireza-mf commented:
Can you give me permission to push updates to the PR?
It's better to have support for `securitySchemes` natively:
```
const spec = tsoas.getOpenApiSpec(typeNamesForSecureTests, {
            components: {
                securitySchemes: {
                    bearerToken: {
                        type: "http",
                        scheme: "bearer",
                        bearerFormat: "JWT",
                        description: "Bearer Token with JWT",
                    },
                    basicAuth: {
                        type: "http",
                        scheme: "basic",
                    },
                }
            }
        });
```
Also added "scope name" support:
```
type EditBookSecureApi = {
    ...
    security: [{ bearerToken: ["book:write", "book:read"] }];
}
```
I've done the changes locally and need permission to update this PR.

chsegala commented:
I added you as collaborator for the repo, let me know if it works!

Christian Segala


On Sat, Jun 8, 2024 at 4:28 PM Alireza ***@***.***> wrote:

> Can you give me permission to push updates to the PR?
> It's better to have support for securitySchemes natively:
>
> const spec = tsoas.getOpenApiSpec(typeNamesForSecureTests, {
>             components: {
>                 securitySchemes: {
>                     bearerToken: {
>                         type: "http",
>                         scheme: "bearer",
>                         bearerFormat: "JWT",
>                         description: "Bearer Token with JWT",
>                     },
>                     basicAuth: {
>                         type: "http",
>                         scheme: "basic",
>                     },
>                 }
>             }
>         });
>
> Also added "scope name" support:
>
> type EditBookSecureApi = {
>     ...
>     security: [{ bearerToken: ["book:write", "book:read"] }];
> }
>
> I've done the changes locally and need permission to update this PR.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ts-oas/ts-oas/pull/8#issuecomment-2156149971>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI5AB2IDRNXUTLKGXMWO2TZGNLPFAVCNFSM6AAAAABI7HBHXKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNJWGE2DSOJXGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

alireza-mf commented:
Worked! Thank you for the contribution I'll publish it in 2 days
